src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs

```diff
@@ -158,7 +158,18 @@ public class JsonInputFormatter : TextInputFormatter
 
                     if (successful)
                     {
-                        return InputFormatterResult.SuccessAsync(model);
+                        if (model == null && !context.AllowEmptyInput)
```

Comment:
> Don't want to disallow the null JSON keyword.

Here, `JsonInputFormatter` is not *disallowing* `null`; it's regarding it as "no input" (like it does an empty request body, or a request body that only contains whitespace). This fits within the general design we've come to, which is that input formatters make their own judgement about what constitutes "input". What then happens with "no input" depends on how you've configured the new global flag.

`JsonInputFormatter` is unique among our built-in input formatters in that it has these surprising cases where it might otherwise supply `null` even when the request body is nonempty. The original issue (#4750) was due to someone getting tripped up by receiving `null` when `ModelState.IsValid` was `true`, so I would have thought a large part of our goal would be to stop the surprising cases happening out-of-the-box. Of course, a developer can make their own input formatter that sometimes returns `null` as "non-empty" if they want, but since our built-in XML and text ones don't, it's much less surprising if the built-in JSON one doesn't either.

If I'm wrong and there are cases where other built-in formatters will return `null` as a "success" result, then I guess more handling is needed, but I'm not aware of that happening.

@rynowak, @Eilon: do you have an opinion on whether stopping the JSON input formatter from supplying `null` alongside `ModelState.IsValid==true` like this does is a desirable result?

> For example, if we do this here, does the JsonPatchInputFormatter also need to change?

`JsonPatchInputFormatter` inherits this from `JsonInputFormatter`.